controllers/imagerepository_controller.go

```diff
@@ -0,0 +1,580 @@
+		if err != nil {
+			log.Error(err, "failed to create pull robot account", "RobotAccountName", pullRobotAccountName, l.Action, l.ActionAdd, l.Audit, "true")
+			return err
+		}
+		if robotAccount == nil {
+			err := fmt.Errorf("unexpected response from Quay: pull robot account data object is nil")
+			log.Error(err, "nil pull robot account")
+			return err
+		}
+
+		err = r.QuayClient.AddPermissionsForRepositoryToRobotAccount(r.QuayOrganization, repository.Name, pullRobotAccount.Name, false)
+		if err != nil {
+			log.Error(err, "failed to add permissions to pull robot account", "RobotAccountName", robotAccountName, l.Action, l.ActionUpdate, l.Audit, "true")
+			return err
+		}
+
+		remoteSecretName := getRemoteSecretName(imageRepository)
+		if err := r.EnsureRemotePullSecret(ctx, imageRepository, remoteSecretName); err != nil {
+			return err
+		}
+
+		if err := r.CreateRemotePullSecretUploadSecret(ctx, pullRobotAccount, imageRepository.Namespace, remoteSecretName, quayImageURL); err != nil {
+			return err
+		}
+
+		status.Credentials.PullRobotAccountName = pullRobotAccountName
+		status.Credentials.PullSecretName = remoteSecretName
+	}
+
+	status.State = imagerepositoryv1beta1.ImageRepositoryStateReady
+	status.Image.URL = quayImageURL
+	status.Image.Visibility = imageRepository.Spec.Image.Visibility
+	status.Credentials.PushRobotAccountName = robotAccountName
+	status.Credentials.PushSecretName = secretName
+	status.Credentials.GenerationTimestamp = &metav1.Time{Time: time.Now()}
+
+	imageRepository.Spec.Image.Name = imageRepositoryName
+	controllerutil.AddFinalizer(imageRepository, ImageRepositoryFinalizer)
+	if err := r.Client.Update(ctx, imageRepository); err != nil {
+		log.Error(err, "failed to update CR after provision")
+		return err
+	} else {
+		log.Info("provisioned image repository and added finalizer")
+	}
+
+	imageRepository.Status = status
+	if err := r.Client.Status().Update(ctx, imageRepository); err != nil {
+		log.Error(err, "failed to update CR status after provision")
+		return err
+	}
+
+	return nil
+}
+
+// RegenerateImageRepositoryCredentials rotates robot account(s) token and updates corresponding secret(s)
+func (r *ImageRepositoryReconciler) RegenerateImageRepositoryCredentials(ctx context.Context, imageRepository *imagerepositoryv1beta1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx)
+
+	quayImageURL := imageRepository.Status.Image.URL
+	robotAccountName := imageRepository.Status.Credentials.PushRobotAccountName
+
+	robotAccount, err := r.QuayClient.RegenerateRobotAccountToken(r.QuayOrganization, robotAccountName)
+	if err != nil {
+		log.Error(err, "failed to refresh push token")
+		return err
+	}
+	secretName := strings.ReplaceAll(robotAccountName, "_", "-")
+	if err := r.EnsureDockerSecret(ctx, imageRepository, robotAccount, secretName, quayImageURL); err != nil {
+		return err
+	}
+	log.Info("Regenerated push token", "RobotAccountName", robotAccountName)
+
+	if isComponentLinked(imageRepository) {
```

Comment:
Tried:
```
apiVersion: appstudio.redhat.com/v1beta1
kind: ImageRepository
metadata:
  name: imagerepository-sample
  labels:
    appstudio.openshift.io/application: test-application
    appstudio.openshift.io/component: devfile-sample-code-with-quarkus
```
But the pull-secret has not been created

Reply:
Thank you, fixed.